swap/simulations_test.go

```diff
@@ -0,0 +1,752 @@
+
+		// get the testService for the creditor
+		peerItem, ok := sim.NodeItem(creditor, bucketKeySwap)
+		if !ok {
+			return errors.New("no swap in simulation bucket")
+		}
+		creditorSvc := peerItem.(*testService)
+
+		for {
+			// let's always be nice and allow a time out to be catched
+			select {
+			case <-ctx.Done():
+				return errors.New("Timed out waiting for all swap peer connections to be established")
+			default:
+			}
+			// the node has all other peers in its peer list
+			if len(debitorSvc.swap.peers) == 1 && len(creditorSvc.swap.peers) == 1 {
+				break
+			}
+			// don't overheat the CPU...
+			time.Sleep(5 * time.Millisecond)
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+		defer cancel()
+
+		// we will send just maxCheques number of cheques
+		maxCheques := 10
+
+		// the peer object used for sending
+		creditorPeer := debitorSvc.peers[creditor]
+
+		// send maxCheques number of cheques
+		for i := 0; i < maxCheques; i++ {
+			// use a price which will trigger a cheque each time
+			creditorPeer.Send(ctx, &testMsgBigPrice{})
+			// we need to wait a bit in order to give time for the cheque to be processed
+			err = waitForChequeProcessed(creditorSvc)
+			if err != nil {
+				return err
+			}
+		}
+
+		// check balances:
+		b1, err := debitorSvc.swap.loadBalance(creditor)
+		if err != nil {
+			return err
+		}
+		b2, err := creditorSvc.swap.loadBalance(debitor)
+		if err != nil {
+			return err
+		}
+
+		if b1 != -b2 {
+			return fmt.Errorf("Expected symmetric balances, but they are not: %d vs %d", b1, b2)
+		}
+		// check cheques
+		var cheque1, cheque2 *Cheque
+		if cheque1, err = debitorSvc.swap.loadLastSentCheque(creditor); err != nil {
+			return errors.New("expected cheques with creditor, but none found")
+		}
+		if cheque2, err = creditorSvc.swap.loadLastReceivedCheque(debitor); err != nil {
+			return errors.New("expected cheques with debitor, but none found")
+		}
+
+		// both cheques (at issuer and beneficiary) should have same cumulative value
+		if cheque1.CumulativePayout != cheque2.CumulativePayout {
+			return fmt.Errorf("Expected symmetric cheques payout, but they are not: %d vs %d", cheque1.CumulativePayout, cheque2.CumulativePayout)
+		}
+
+		// check also the actual expected amount
+		expectedPayout := uint64(maxCheques * (DefaultPaymentThreshold + 1))
+
+		if cheque2.CumulativePayout != expectedPayout {
+			return fmt.Errorf("Expected %d in cumulative payout, got %d", expectedPayout, cheque1.CumulativePayout)
+		}
+
+		return nil
+
+	})
+
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+
+	log.Info("Simulation ended")
+}
+
+// TestBasicSwapSimulation starts 16 nodes, then in a simple round robin fashion sends messages to each other.
+// Then checks that accounting is ok. It checks the actual amount of balances without any cheques sent,
+// in order to verify that the most basic accounting works.
+func TestBasicSwapSimulation(t *testing.T) {
```

Comment:
can you explain why this test is needed? you have unit tests and integration tests with pretty good coverage. are you trying to test if `devp2p` works correctly or?

Reply:
It is tricky to answer this. Yes, we have pretty good test coverage. Yes, the test is maybe not totally needed. However from my integration test experience, you always want to make sure that all the pieces actually work well together. 

This test makes sure that the actual accounting over devp2p works *with the Swap* objects and **without any external dependency** like any of the `stream` or `retrieval` protocols. It is the best we can do as an integration test without having to build a complex test with all needed pieces (blockchain, oracles, accounted protocols) which we would have to do as a deployment on k8s (like the smoke tests).

While it's usefulness is debatable, I still believe it has merit to have it and I would like to keep it.

Reply:
I understand. And although I agree with all of the above you've mentioned I think we should be careful with the amounts of tests we introduce to the test suite mainly because simulations are very difficult to debug, especially ones that are larger than 2-3 nodes. 
Also, for this particular test I have a feeling that the test case is already covered in the previous simulation and that the unit tests in swap cover this too (without all of the plumbing around). So I am wary of adding another 200 lines of code for something which is already tested in previous test cases.
Personally I feel there's a lot of code in this test and I don't feel it gives much added value over the other test cases.
If you wanna test the protocol infrastructure you can easily add a short integration test that tests if the protocol is initialized correctly, if you want to test e2e that accounting works with the protocol that's fine but I suggest you coalesce the tests to one or maximum two test cases that check that the accounting works (why do you need a 2 node test when you have a 16 node test and vice versa? both overlap and test for the same thing which is operating so i'm not sure about the necessity of both of the test cases).
Another note about more tests is that as much as I love integration test coverage I think our smoke tests unravel much more information about the performance and correctness of our codebase than endless simulations which are difficult to debug, and if you write a smoke test that would regularly run you eliminate the need to test all of this protocol boilerplating because if the protocol doesn't work right your smoke test will catch it immediately.
2 cents, not more.

Reply:
I agree with you generally. Just a couple of comments:
* Although most elements are indeed tested with existing tests, however not in such a "e2e" way. The accounting happens at `p2p/protocols` level, this test is at a higher level through the `swap` package. 
* I am particularly proud of this simulation because I am 100% sure this isn't flaky (the 16 nodes one, because it doesn't use the `SimulatedBackend` and is totally under control. The other 2 nodes onese have been flaky sometimes). It actually is a show-case simulation on how to look for sync points to eliminate "flakyness". 
* There is a historical reason for this sim to exist: Initially we planned to have many configurations tested with in-proc sims, and thus this one was created as a first / pilot (thus the branch name `incentives-first-sim`). Later we realized that in-proc sims may not be the best tools due to the many dependencies of swap now (blockchain, oracles, etc.). At this point though this simulation had already been implemented. The factual arguments are a bit thin, and I am mainly clinging on this simulation because of the work which went into it. 
* >why do you need a 2 node test when you have a 16 node test and vice versa?
  They don't overlap. The 16 nodes simulation (`BasicSimulation`) only tests basic accounting (no cheques). The other 2 node simulation don't test accounting at all but are focused on repeated checks being issued and verify that that does not mess up cheque history / values.
* The last comment on integration tests is absolutely accurate. Indeed we will (have to) write more such integration tests because there are moving parts we need to address (again, blockchains, oracles, etc difficult to integrate  with the in-proc setup). So these simulations in this PR, being short and run as part of the go test suite, will be a first checkpoint that things are ok before having to deploy clusters to check on the correctness of things.

Now I have enumerated all my arguments. Again, mostly attached to this PR due to the amount of work which went into it. If all the above arguments (and others mentioned elsewhere in the PR) are fully understood, I will finally not resist eliminating some or all of its content if reviewers come to the conclusion that they are irrelevant / obsolete. Thank you.

Reply:
> The other 2 node simulation don't test accounting at all but are focused on repeated checks being issued and verify that that does not mess up cheque history / values.

I am still oblivious on why you would need so many simulations for this. I understand that a lot of work went into this and I respect that, but from what you're saying this could also probably be reworked to be a single table driven test with different payment thresholds set on every run, that way you test that both ping-pong accounting works with and without cheques. The usage of multi-node simulations in SWAP IMHO does not bring much added value since the component is decoupled from kademlia notions and there aren't any paid service delegation to other nodes (i.e. node A cannot pay node B to ask node C for a certain content and in this scenario a cluster does not bring any added value).